README.markdown
# Codem-transcode

* http://github.com/madebyhiro/codem-transcode

## Description

Codem-transcode is an offline video transcoder written in node.js. It

1. Uses ffmpeg for transcoding
2. Has a simple HTTP API
3. Is mostly asynchronous

Codem-transcode can be used in conjunction with Codem-schedule (https://github.com/madebyhiro/codem-schedule) for robust job scheduling and notifications or it can be used stand-alone, with or without a custom scheduler.

## Requirements

* ffmpeg (at least 0.10 and compiled/configured to your own taste)
* sqlite3/MySQL/PostgreSQL
* node.js version 0.8.x (x>=11), with packages (if you use npm they will be installed automatically):
    * sequelize (http://sequelizejs.com/)
    * sqlite3 (http://github.com/developmentseed/node-sqlite3)
    * express (http://expressjs.com/)
    * argsparser (http://github.com/kof/node-argsparser)
    * mkdirp (https://github.com/substack/node-mkdirp)
    * async (https://github.com/caolan/async)

## Installation

The recommended installation procedure is to just use `npm` (http://npmjs.org/):

    # npm install codem-transcode

Install it to your preferred location, or use the `-g` option to install it globally.

## Upgrading

Upgrading should most of the times be as simple as shutting down, installing the new package and restarting the transcoder. Unless you're...

### Upgrading from earlier versions to 0.5 (IMPORTANT!)

Codem-transcode switched from using "plain" SQL (using sqlite) to a database abstraction layer (Sequelize). This brings some advantages
(support for multiple database engines, better consistency, easier migrations in the future), but is not backwards compatible. Therefore, we recommend you backup and move away your old database and
start with a fresh one. The procedure for this would be:

*   Shutdown the transcoder;
*   Move your database away (or delete if you're not interested in the history);
*   Install the new package;
*   Start the transcoder.

This will initialize a new up-to-date database which can be migrated to newer schema's more easily in the future. If you are doing a
clean install you do not need to worry about any of this.

If you want to keep your history we recommend you follow the above procedure and write a separate import script to import your old data
into the new database.

## Starting

When you install codem-transcode a script will be installed that allows you to start the transcoder. If you install it globally it should be in your `PATH`, otherwise, you can start the transcoder using:

    # /PATH/TO/TRANSCODER/bin/codem-transcode
    
Please check for yourself where `npm` installs your packages and script.

## Configuration

Configuration is done by specifying a CLI option (`-c`) and pointing to a file containing a valid JSON object (http://www.json.org/). Note that node.js' JSON parser is fairly strict so make sure you get the syntax right. An example config is:

    {
        "port":            8080,
        "access_log":      "/var/log/access_log",
        "database":        "/var/db/jobs.db",
        "slots":           8,
        "interface":       "127.0.0.1",
        "encoder":         "ffmpeg",
        "scratch_dir":     "/tmp",
        "use_scratch_dir": true,
        "ffprobe":         null
    }

Configuration options:

* `port`; port to start server on, default `8080`
* `interface`; which network interface to listen on, default `127.0.0.1` (only `localhost`)
* `access_log`; location to store HTTP access log, default `/var/log/access_log`
* `database`; location to store the jobs database, default is SQLite with `/var/db/jobs.db`
* `slots`; number of transcoding slots to use (i.e. the maximum number of ffmpeg child processes), defaults to the number of CPUs/cores in your machine
* `encoder`; path to the ffmpeg binary, if it is in your path specifying only `ffmpeg` is sufficient, defaults to `ffmpeg`
* `scratch_dir`; temporary files are written here and moved into the destination directory after transcoding, defaults to `/tmp`
* `use_scratch_dir`; if set to false temporary files will be written to the output directory of your job, for setups that don't require or are not able to use a separate `scratch_dir`. Defaults to `true` so if you don't want to disable the `scratch_dir` you can also omit this option from your config file.
* `ffprobe`; path to the ffprobe binary, if it is in your path specifying only `ffprobe` is sufficient, defaults to `null`. Set this to a non-null value if you want to enable ffprobe support in the transcoder.

Note that the default config will put the access_log and job database in `/var/log` and `var/db/` respectively. If you wish to put these in a different location please supply your own config. You can start the transcoder with your custom config using:

    # /PATH/TO/TRANSCODER/bin/codem-transcode -c /PATH/TO/CONFIG/config.json

### Advanced database configuration

codem-transcode supports multiple database backends, courtesy of Sequelize. The default is still to store data in a SQLite database (whenever you specify a string for `database` in the config file). To use MySQL or Postgres, supply a valid object for the database entry. Your configuration will then look like:

    {
        "port":            8080,
        "access_log":      "/var/log/access_log",
        "database":        {
            "dialect": "mysql",
            "username": "root",        
            "database": "codem",
            "host": "localhost",
            "port": 3306
        },
        "slots":           8,
        "interface":       "127.0.0.1",
        "encoder":         "ffmpeg",
        "scratch_dir":     "/tmp",
        "use_scratch_dir": true,
        "ffprobe":         null
    }

Be sure to specify a `dialect` ("mysql", "postgres", "sqlite"), a `username`, a `password` (can be omitted if using a passwordless database) and a `host` (can be omitted for "localhost"). `port` can be omitted for the default port.

## Usage

After starting the server you can control it using most HTTP CLI tools, such as `curl` or `wget`. The HTTP API is as follows:

* * *
Request: `POST /jobs`

Parameters (HTTP POST data, should be valid JSON object):

    {
        "source_file": "/PATH/TO/INPUT/FILE.wmv",
        "destination_file":"/PATH/TO/OUTPUT/FILE.mp4",
        "encoder_options": "-acodec libfaac -ab 96k -ar 44100 -vcodec libx264 -vb 416k -s 320x180 -y -threads 0",
        "thumbnail_options": {
            "percentages": [0.25, 0.5, 0.75],
            "size": "160x90",
            "format": "png"
        },
        "segments_options" : {
          "segment_time": 10
        },
        "callback_urls": ["http://example.com/notifications"]
    }

Responses:

* `202 Accepted` - Job accepted
* `400 Bad Request` - Invalid request (format)
* `503 Service Unavailable` - Transcoder not accepting jobs at the moment (all encoding slots are in use)


Required options are `source_file`, `destination_file` and `encoder_options`. Input and output files must be *absolute* paths.

The `callback_urls` array is optional and is a list (array) of HTTP endpoints that should be notified once encoding finishes (due to the job being complete or some error condition). The notification will sent using HTTP PUT to the specified endpoints with the job status. It will also include a custom HTTP header "X-Codem-Notify-Timestamp" that contains the timestamp (in milliseconds) at which the notification was generated and sent. It is best to observe this header to determine the order in which notifications are received at the other end due to network lag or other circumstances that may cause notifications to be received out of order.

The `thumbnail_options` object is optional and contains a set of thumbnails that should be encoded after the transcoding is complete. Thumbnails are captured from the source file for maximum quality. The options for thumbnails include:

* Either "percentages" or "seconds" (but not both at the same time), valid options are:
    * A single percentage, this will trigger a thumbnail every x%. `"percentages": 0.1` will generate thumbnails at 0%, 10%, 20%, [...], 100%.
    * An array of explicit percentages, this will trigger thumbnails only at the specified positions.
      `"percentages": [0.25, 0.5, 0.75]` will generate thumbnails at 25%, 50% and 75%.
    * A single offset in seconds, this will trigger a thumbnail every x seconds. `"seconds": 10` will generate thumbnails at 0 seconds, 10 seconds, 20 seconds, etc., until the end of the source file.
    * An array of explicit offsets, this will trigger thumbnails only at the specified positions.
      `"seconds": [30, 60, 90]` will generate thumbnails at 30 seconds, 60 seconds and 90 seconds.
* A size can be specified in pixels (width x height). If omitted it will generate thumbnails the size of the source video. (optional)
* A format for the thumbnails. The format must be supported by your ffmpeg binary. If omitted it will generate thumbnails in the JPEG format. Most people will use either "jpg" or "png". (optional)

If you specify thumbnails but an error occurs during generation, your job will be marked as failed. If you don't specify a valid `seconds` or `percentages` option thumbnail generation will be skipped but the job can still be completed successfully.

The `segments_options` object is optional and contains segment time (duration) in seconds. Segmented videos are used in [HLS](https://en.wikipedia.org/wiki/HTTP_Live_Streaming). These options are applied to the encoded video, thus `encoder_options` are required. Moreover `encoder_options` should prepare video for segmenting, because bitstream
filter [h264_mp4toannexb](https://www.ffmpeg.org/ffmpeg-bitstream-filters.html#h264_005fmp4toannexb) will be applied to the video. Therefore it is recommended to transcode to an MP4 file before segmenting.

The segmenting command looks like:

    ffmpeg -i /tmp/46ee0a404a4b75d85c09d98a7c6b403579ee9f99.mp4 -codec copy -map 0 \
      -f segment -vbsf h264_mp4toannexb -flags -global_header -segment_format mpegts \
      -segment_list /path/to/dest_file_dir/dest_file_name.m3u8 -segment_time 10 \
      /path/to/dest_file_dir/dest_file_name-%06d.ts

`46ee0a404a4b75d85c09d98a7c6b403579ee9f99.mp4` is a temporary encoded file (generated by Codem). After transcoding and segmenting you end up with the transcoded file, as well as the segments/playlist.

### Thumbnail-only job

It's possible to only generate thumbnails from a video and not do any transcoding at all. This might come in handy if you're transcoding to lots of different formats and want to keep thumbnail generation separate from transcoding. You achieve this by POSTing a job with `"encoder_options": ""` (empty string) and of course specifying your `thumbnail_options`. In this case `destination_file` should be a _prefix_ for the output file, e.g. `"destination_file": "/Users/codem/output/my_video"` results in thumbnails in `/Users/codem/output/` with filenames such as `my_video-$offset.$format` (where `$offset` is the thumbnail offset in the video and `$format` of course the thumbnail format). All other options remain the same. See the examples.

* * *
Request: `GET /jobs`

Responses:

* `200 OK` - Returns status of all active jobs

* * *
Request: `GET /jobs/$JOB_ID`

Responses:

* `200 OK` - Returns status of job
* `404 Not Found` - Job not found

* * *
Request: `DELETE /jobs/$JOB_ID`

Cancels the job (if it is running) and deletes it from the database.

Responses:

* `200 OK` - Returns last known status of the job that is being deleted
* `404 Not Found` - Job not found

* * *
Request: `DELETE /jobs/purge/$AGE`

Purge successfully completed jobs from the database with a certain age. Age is specified in seconds since it was created. So an age of 3600 deletes jobs that were successful and created more than 1 hour ago.

Responses:

* `200 OK` - Returns number of jobs that were purged.

* * *
Request: `POST /probe`

Probe a source file using `ffprobe` (if you have enabled it in the configuration). Output is a JSON object containing the `ffprobe` output.

Parameters (HTTP POST data, should be valid JSON object):

    {
        "source_file": "/PATH/TO/INPUT/FILE.wmv"
    }

Responses:

* `200 OK` - Returns `ffprobe` output JSON-formatted
* `400 Bad Request` - Returned if you attempt to probe a file when there is no path set to the `ffprobe` binary
* `500 Internal Server Error` - Returned if there was an error while trying to probe, the output from `ffprobe` will be returned as well

* * *
## Examples

Create a new job, transcode "video.wmv" to "video.mp4" using the specified ffmpeg options (96kbit/s audio, 416kbit/s video, 320x180, use as much threads as possible). Requires libx264 support in your ffmpeg.

    # curl -d '{"source_file": "/tmp/video.wmv","destination_file":"/tmp/video.mp4","encoder_options": "-acodec libfaac -ab 96k -ar 44100 -vcodec libx264 -vb 416k -s 320x180 -y -threads 0"}' http://localhost:8080/jobs

    Output: {"message":"The transcoder accepted your job.","job_id":"d4b1dfebe6860839b2c21b70f35938d870011682"}
    

CHANGELOG.markdown
## codem-transcode 0.5.11 (2015/12/02) ##

* Sanitize null values (@brain64bit)
* Allow HTTPS for notification delivery (@mirion)

## codem-transcode 0.5.10 (2015/09/16) ##

* Increase maxBuffer to 4MB to compensate for older ffmpegs

## codem-transcode 0.5.9 (2015/09/03) ##

* Fix for lingering "processing" jobs. (#42)
* Encoder options can be an array instead of a string as well (to allow complex filters)

## codem-transcode 0.5.8 (2015/07/05) ##

* Prevent transcoder from returning a negative freeSlot count (@larsfi).

## codem-transcode 0.5.7 (2015/04/23) ##

* Add option for database port in config (@brain64bit).

## codem-transcode 0.5.6 (2015/04/22) ##

* Return a job from memory if it is still transcoding.

## codem-transcode 0.5.5 (2014/09/08) ##

* Added basic HLS support (@cutalion) using "segments_options".

## codem-transcode 0.5.4 (2014/03/18) ##

* Updated sqlite3 dependency to 2.2.0.

## codem-transcode 0.5.3 (2014/01/27) ##

* Bugfix: issue #19 and #20, missing or corrupt ffmpeg will no longer put transcoder in weird state.
* Send "X-Codem-Notify-Timestamp" header for notifications with millisecond precision (instead of second).

## codem-transcode 0.5.2 (2013/09/10) ##

* Bugfix: migrations should create "Jobs" table (instead of "jobs")

## codem-transcode 0.5.1 (2013/07/08) ##

*   Extra path separator / nul file check for platform-specific issues (Windows/*nix)
*   Allow thumbnail-only jobs. See README.

## codem-transcode 0.5.0 (2013/07/02) ##

*   All fixes/features from the betas.
*   Use async package for some of the complex callback chains.
*   Ability to capture thumbnails from the videos after transcoding.

## codem-transcode 0.5.0-beta.4 (2013/04/30) ##

*   Fix for #15, cannot open database when using default config.
*   sqlite3 dependency updated to 2.1.7.

## codem-transcode 0.5.0-beta.3 (2013/04/25) ##

*   Allow purging of old successful jobs. See README for usage.
*   Tested against Node.js 0.10.

## codem-transcode 0.5.0-beta.2 (2013/03/07) ##

*   Switch to Sequelize for database abstraction. codem-transcode now supports SQLite, MySQL and Postgres. PLEASE NOTE
    that this change is *not* backwards compatible. You will need to update your database. If you are coming from an old
    version of codem-transcode we advise you to move away your old database and let the software generate a new database
    for you.
    
## codem-transcode 0.5.0-beta.1 (2013/02/19) ##

*   The logging system should now be more able to handle unexpected issues, such as when your logging file system is out
    of space. In most cases the system should be able to gracefully intercept any issues and resume operation after the
    issue has been cleared (ie. more space has been created).

## codem-transcode 0.4.4 (2013/01/28) ##

*   Use a "=" instead of a "LIKE" when loading a job. Prevents excessive disk IO.

## codem-transcode 0.4.3 (2013/01/22) ##

*   Allow both time formats in the ffmpeg output parsing (xx:xx and xx:xx:xx).

## codem-transcode 0.4.2 (2013/01/07) ##

*   Fix for copying a file across partitions.

## codem-transcode 0.4.1 (2012/10/16) ##

*   Added ffprobe support.

## codem-transcode 0.4.0 (2012/10/10) ##

*   Updated package to Node 0.8.

## codem-transcode 0.3.1 (2012/06/27) ##

*   Allow disabling of scratch directory.

## codem-transcode 0.3.0 (2012/03/20) ##

*   Updated dependency to a newer version of Node.
*   Update ffmpeg time parsing.
*   Switched from connect to express.

## codem-transcode 0.2.1 (2011/09/26) ##

*   Add additional "X-Codem-Notify-Timestamp" HTTP header.

## codem-transcode 0.2.0 (2011/08/23) ##

*   Deleting and cancelling jobs.
*   Additional error checking.

## codem-transcode 0.1.2 (2011/07/11) ##

*   Additional logging.
*   Notify when duration is known.
*   Added extra check when creating directories.

## codem-transcode 0.1.1 (2011/05/24) ##

*   Fixed license info

## codem-transcode 0.1.0 (2011/05/24) ##

*   Initial release

lib/job.js
var util          = require('util'),
    crypto        = require('crypto'),
    fs            = require('fs'),
    child_process = require('child_process'),
    config        = require('./config').load()
    mkdirp        = require('mkdirp'),
    path          = require('path'),
    notifyHandler = require('./notify-handler'),
    logger        = require('./logger'),
    Sequelize     = require('sequelize'),
    async         = require('async'),
    os            = require('os');

var StatusCodes = {
  SUCCESS: "success",
  FAILED: "failed",
  PROCESSING: "processing"
}

var JobUtils = {
  sql: null,

  generateThumbnailPath: function(destinationFile, offset, format) {
    return [path.dirname(destinationFile), path.basename(destinationFile, path.extname(destinationFile))].join(path.sep) + "-" + offset + "." + format;
  },

  generateRangeFromThumbOpts: function(thumbOpts, duration) {
    if (thumbOpts['percentages']) {
      // percentage based thumbnails
      return JobUtils.percentagesToRange(thumbOpts['percentages'], duration);
    } else if (thumbOpts['seconds']) {
      // seconds based thumbnails
      return JobUtils.secondsToRange(thumbOpts['seconds'], duration);
    } else {
      return null;
    }
  },

  getDatabase: function() {
    if (JobUtils.sql == null) {
      if (config['database']['dialect'] == "sqlite") {
        JobUtils.sql = new Sequelize('database', 'username', 'password', {
          dialect: 'sqlite',
          storage: config['database']['database'],
          logging: false
        });
      } else {
        JobUtils.sql = new Sequelize(config['database']['database'], config['database']['username'], config['database']['password'], {
          dialect: config['database']['dialect'],
          storage: config['database']['database'],
          host: config['database']['host'],
          port: config['database']['port'],
          omitNull: true,
          pool: false,
          logging: false
        });
      }
    }
    return JobUtils.sql;
  },

  getMask: function() {
    return JobUtils.pad((process.umask() ^ 0777).toString(8), '0', 4);
  },

  markUnknownState: function(callback) {
    Job.findAll({ where: { status: StatusCodes.PROCESSING }}).success(function(result) {
      if (result.length > 0) {
        for (var job in result) {
          result[job].status = StatusCodes.FAILED;
          result[job].message = "The transcoder quit unexpectedly or the database was unavailable for some period.";
          result[job].save();
        }
      }
    });
    callback(null);
  },

  migrateDatabase: function(callback) {
    var migrator = JobUtils.getDatabase().getMigrator({path: __dirname + "/../migrations" });
    migrator.migrate().success(function() {
      callback(null);
    }).error(function(err) {
      callback(err);
    });
  },

  pad: function(orig, padString, length) {
    var str = orig;
    while (str.length < length)
        str = padString + str;
    return str;
  },

  percentagesToRange: function(percentages, duration) {
    if (Array.isArray(percentages)) {
      // explicit percentages
      return percentages.map(function (p) { return Math.floor(p * duration); });
    } else {
      // single percentage interval
      var offsets = [];
      for (var p = 0.0; p <= 1.0; p += percentages) {
        offsets.push(Math.floor(p * duration));
      }
      return offsets;
    }
  },

  secondsToRange: function(seconds, duration) {
    if (Array.isArray(seconds)) {
      // explicit percentages
      return seconds;
    } else {
      // single percentage interval
      var offsets = [];
      for (var offset = 0; offset <= duration; offset += seconds) {
        offsets.push(Math.floor(offset));
      }
      return offsets;
    }
  },

  verifyDatabase: function(callback) {
    JobUtils.getDatabase().getQueryInterface().showAllTables().success(function (tables) {
      if (tables.length > 0 && tables.indexOf('SequelizeMeta') == -1) {
        logger.log("You appear to be upgrading from an old version of codem-transcode (<0.5). The database handling has " +
                   "changed, please refer to the upgrade instructions. To prevent data loss the transcoder will now exit.");
        callback("Old database schema detected.");
      } else {
        callback(null);
      }
    }).error(function (err) {
      logger.log(err);
      callback(err);
    });
  }
}

// Model definition
var Job = JobUtils.getDatabase().define('Job', {
  id:          { type: Sequelize.INTEGER, primaryKey: true, autoIncrement: true },
  internalId:  { type: Sequelize.STRING, defaultValue: null },
  status:      { type: Sequelize.STRING, defaultValue: StatusCodes.PROCESSING },
  progress:    { type: Sequelize.FLOAT, defaultValue: 0.0 },
  duration:    { type: Sequelize.INTEGER, defaultValue: 0 },
  filesize:    { type: Sequelize.INTEGER, defaultValue: 0 },
  opts:        { type: Sequelize.TEXT, defaultValue: null },
  thumbnails:  { type: Sequelize.TEXT, defaultValue: null },
  message:     { type: Sequelize.TEXT, defaultValue: null },
  playlist:    { type: Sequelize.STRING, defaultValue: null },
  segments:    { type: Sequelize.TEXT, defaultValue: null },
  createdAt:   Sequelize.DATE,
  updatedAt:   Sequelize.DATE
}, {
  classMethods: {
    prepareDatabase: function(callback) {
      JobUtils.verifyDatabase(function (err) {
        if (err) {
          callback(err);
        } else {
          JobUtils.migrateDatabase(function (err) {
            if (err) {
              callback(err);
            } else {
              JobUtils.markUnknownState(function (err) {
                err ? callback(err) : callback(null);
              })
            }
          });
        }
      })
    },
    create: function(opts, callback, completeCallback) {
      var job = Job.build({ opts: JSON.stringify(opts), internalId: Job.generateId() });
      job.save().success(function(job) {
        job.prepare(function onPrepared(message) {
          if (message == "hasInput")     job.hasInput     = true;
          if (message == "hasOutputDir") job.hasOutputDir = true;
          if (job.hasInput && job.hasOutputDir) job.spawn();
        })
        job.completeCallback = completeCallback;
        callback(null, job);
      }).error(function(error) {
        /*
          Error while saving job, this appears to not always be called(!).
          Known issue in either sequelize or mysql package.
        */
        logger.log("Could not write job " + job.internalId + " to the database.");
        callback('Unable to save new job to the database', null);
      });
    },
    generateId: function() {
      var hash = crypto.createHash('sha1');
      var date = new Date();

      hash.update([date, Math.random()].join(''));

      return hash.digest('hex');
    }
  },
  instanceMethods: {
    parsedOpts: function() {
      return JSON.parse(this.opts);
    },

    prepare: function(callback) {
      var job = this;
      var destination_dir = path.dirname(path.normalize(job.parsedOpts()['destination_file']));
      var mode = JobUtils.getMask();

      fs.stat(job.parsedOpts()['source_file'], function(err, stats) {
        if (err) {
          job.exitHandler(-1, "unable to read input file (" + job.parsedOpts()['source_file'] + ").");
        } else {
          if (stats.isFile()) { job.filesize = stats.size || Number.NaN };
          callback("hasInput");
        }
      });

      mkdirp.mkdirp(destination_dir, mode, function(err) {
        if (err) {
          fs.stat(destination_dir, function(err, stats) {
            if (err) {
              job.exitHandler(-1, "unable to create output directory (" + destination_dir + ").");
            } else {
              if (stats.isDirectory()) {
                callback("hasOutputDir");
              } else {
                job.exitHandler(-1, "unable to create output directory (" + destination_dir + ").");
              }
            }
          });
        } else {
          callback("hasOutputDir");
        }
      });

    },
    spawn: function() {
      var job = this;
      if (this.hasInput && this.hasOutputDir && !this.the_process) {
        var args = [];
        args.push('-i', this.parsedOpts()['source_file']);

        var extension = path.extname(this.parsedOpts()['destination_file']);
        var outputDir = path.dirname(this.parsedOpts()['destination_file']);

        if (this.parsedOpts()['encoder_options'].length > 0) {
          // "proper" transcoding job
          if (Array.isArray(this.parsedOpts()['encoder_options'])) {
            args = args.concat(this.parsedOpts()['encoder_options']);
          } else {
            args = args.concat(this.parsedOpts()['encoder_options'].replace(/\s+/g, " ").split(' '));
          }
          var tmpFile = outputDir + path.sep + this.internalId + extension;

          if (config['use_scratch_dir'] == true) {
            tmpFile = config['scratch_dir'] + path.sep + this.internalId + extension;
          }

          args.push(tmpFile);

          job.tmpFile = tmpFile;
        } else {
          // thumbnail only job, but still need to find duration, so we start a "null" job
          var null_file = (!!os.platform().match(/^win/) ? 'nul' : '/dev/null');
          args.push('-f', 'null', '-acodec', 'copy', '-vcodec', 'copy', '-y', null_file);
        }

        var the_process = child_process.spawn(config['encoder'], args);

        the_process.stderr.on('data', function(data) { job.progressHandler(data); });
        the_process.on('error', function(err) { job.lastMessage = 'Unable to execute the ffmpeg binary: ' + err; job.didFinish(1); });
        the_process.on('exit', function(code) { job.didFinish(code); });

        this.the_process = the_process;
      }
    },
    cancel: function() {
      if (this.the_process) {
        this.the_process.kill();
        this.exitHandler(-1, 'job was cancelled');
      }
    },
    didFinish: function(code) {
      if (code != 0) {
        this.finalize(code);
        return;
      }

      this.processThumbnails({
        error:   function(job) { job.finalize(1); },
        success: function(job) {
          job.processSegments({
            error:   function(job) { job.finalize(1); },
            success: function(job) { job.finalize(0); }
          })
        }
      })
    },

    processSegments: function(callbacks){
      if (!this.parsedOpts()['segments_options']) {
        callbacks.success(this);
        return;
      }

      logger.log("Processing segments for job " + this.internalId + ".");

      var job              = this;
      var args             = [];
      var segmentsOpts     = this.parsedOpts()['segments_options'];
      var segmentTime      = segmentsOpts['segment_time'];
      var destinationFile  = this.parsedOpts()['destination_file'];
      var playlistName     = path.basename(destinationFile, path.extname(destinationFile));
      var playlistDir      = path.dirname(destinationFile);
      var playlistPath     = [playlistDir, playlistName].join(path.sep) + '.m3u8';
      var segmentsFormat   = [playlistDir, playlistName].join(path.sep) + '-%06d.ts'
      var inputFile        = (this.parsedOpts()['encoder_options'].length > 0) ? job.tmpFile : this.parsedOpts()['source_file']

      args.push('-i', inputFile,
                '-codec', 'copy', '-map', '0', '-f', 'segment',
                '-vbsf', 'h264_mp4toannexb', '-flags', '-global_header',
                '-segment_format', 'mpegts', '-segment_list', playlistPath,
                '-segment_time', segmentTime, segmentsFormat);

      child_process.execFile(config['encoder'], args, { maxBuffer: 4096*1024 }, function(error, stdout, stderr) {
        if (error) {
          job.lastMessage = 'Error while generating segments: ' + error.message;
          callbacks.error(job);
          return;
        }

        fs.readdir(playlistDir, function(error, files) {
          if (error) {
            job.lastMessage = 'Error while generating segments: ' + error.message;
            callbacks.error(job);
            return;
          }

          job.segments = JSON.stringify(files.filter(
            function(file){ return file.match(new RegExp(playlistName + "-\\d+\\.ts")) }
          ).map(
            function(file){ return path.join(playlistDir, file) }
          ));

          job.playlist = playlistPath;

          callbacks.success(job);
        });
      });
    },

    processThumbnails: function(callbacks) {
      if (!this.parsedOpts()['thumbnail_options']) {
        callbacks.success(this);
        return;
      }

      logger.log("Processing thumbnails for job " + this.internalId + ".");
      var thumbOpts = this.parsedOpts()['thumbnail_options'];
      var range = JobUtils.generateRangeFromThumbOpts(thumbOpts, this.duration);

      if (!range) {
        // no valid range
        logger.log("No valid thumbnails to process for job " + this.internalId + ". Skipping...");
        callbacks.success(this);
        return;
      }

      var job = this;
      async.parallel(
        range.map(job.execThumbJob.bind(job)),
        function(err, results) {
          if (err) {
            job.lastMessage = err.message;
            callbacks.error(job);
          } else {
            job.thumbnails = JSON.stringify(results);
            callbacks.success(job);
          }
        }
      );
    },
    execThumbJob: function(offset) {
      var job = this;
      return function(callback) {
        var thumbOpts = job.parsedOpts()['thumbnail_options'];
        var args = ['-ss', offset, '-i', job.parsedOpts()['source_file'], '-vframes', '1', '-y'];

        // Explicit size provided
        if (thumbOpts['size'] && thumbOpts['size'] != 'src') {
          args.push('-s', thumbOpts['size']);
        }

        var format = thumbOpts['format'] ? thumbOpts['format'] : 'jpg';
        var destinationFile = job.parsedOpts()['destination_file'];

        // Destination file + offset + extension
        var outputFile = JobUtils.generateThumbnailPath(destinationFile, offset, format);
        args.push(outputFile);

        var thumb_process = child_process.execFile(config['encoder'], args, { maxBuffer: 4096*1024 }, function(error, stdout, stderr) {
          if (error) {
            callback(new Error('Error while generating thumbnail: ' + error.message), null);
          } else {
            callback(null, outputFile);
          }
        });
      }
    },
    finalize: function(code) {
      var job = this;

      if (code != 0) {
        job.exitHandler(code, "ffmpeg finished with an error: '" + job.lastMessage + "' (" + code + ").");
        return;
      }

      if (!job.tmpFile) {
        // No tmpFile, hence no transcoding, only thumbnails or segmenting
        if (job.parsedOpts()['thumbnail_options']) {
          job.exitHandler(code, 'finished thumbnail job.');
        } else {
          job.exitHandler(code, 'finished segmenting job.');
        }
        return;
      }

      fs.rename(job.tmpFile, job.parsedOpts()['destination_file'], function (err) {
        if (err) {
          if ( (err.message).match(/EXDEV/) ) {
            /*
              EXDEV fix, since util.pump is deprecated, using stream.pipe
              example from http://stackoverflow.com/questions/11293857/fastest-way-to-copy-file-in-node-js
            */
            try {
              logger.log('ffmpeg finished successfully, trying to copy across partitions');
              fs.createReadStream(job.tmpFile).pipe(fs.createWriteStream(job.parsedOpts()['destination_file']));
              job.exitHandler(code, 'ffmpeg finished succesfully.');
            } catch (err) {
              logger.log(err);
              job.exitHandler(-1, 'ffmpeg finished succesfully, but unable to move file to different partition (' + job.parsedOpts()['destination_file'] + ').');
            }

          } else {
            logger.log(err);
            job.exitHandler(-1, 'ffmpeg finished succesfully, but unable to move file to destination (' + job.parsedOpts()['destination_file'] + ').');
          }
        } else {
          job.exitHandler(code, 'ffmpeg finished succesfully.');
        }
      });
    },
    toJSON: function() {
      var obj = {
        'id': this.internalId,
        'status': this.status,
        'progress': this.progress,
        'duration': this.duration,
        'filesize': this.filesize,
        'message': this.message,
      };

      if (this.thumbnails) {
        obj['thumbnails'] = JSON.parse(this.thumbnails);
      }

      if (this.playlist) { obj['playlist'] = this.playlist; }
      if (this.segments) {
        obj['segments'] = JSON.parse(this.segments);
      }

      return obj;
    },
    progressHandler: function(data) {
      if (this.hasExited) return;

      this.lastMessage = data.toString().replace("\n",'');

      (isNaN(this.duration) || this.duration == 0) ? this.extractDuration(data.toString()) : this.extractProgress(data.toString());

      this.save().success(function() {
        // successfull save
      }).error(function(err) {
        // error while saving job
      });
    },
    extractDuration: function(text) {
      if (!this.durationBuffer) this.durationBuffer = "";

      this.durationBuffer += text;
      var re = new RegExp(/Duration:\s+(\d{2}):(\d{2}):(\d{2}).(\d{1,2})/);
      var m = re.exec(this.durationBuffer);

      if (m != null) {
        var hours = parseInt(m[1], 10), minutes = parseInt(m[2], 10), seconds = parseInt(m[3], 10);

        this.duration = hours * 3600 + minutes * 60 + seconds;
        notifyHandler.notify(this);
      }
    },
    extractProgress: function(text) {
      // 00:00:00 (hours, minutes, seconds)
      var re = new RegExp(/time=(\d{2}):(\d{2}):(\d{2})/);
      var m = re.exec(text);

      if (m != null) {
        var hours = parseInt(m[1], 10), minutes = parseInt(m[2], 10), seconds = parseInt(m[3], 10);
        var current = hours * 3600 + minutes * 60 + seconds;
        this.progress = current / this.duration;
      } else {
        // 00.00 (seconds, hundreds)
        re = new RegExp(/time=(\d+).(\d{2})/);
        m = re.exec(text);

        if (m != null) {
          var current = parseInt(m[1], 10);
          this.progress = current / this.duration;
        }
      }
    },
    exitHandler: function(code, message) {
      this.hasExited = true
      this.message = message;

      if (code == 0) {
        this.progress = 1.0;
      }

      this.status = (code == 0 ? StatusCodes.SUCCESS : StatusCodes.FAILED);
      this.save().success(function() {
        // successfull save
      }).error(function(err) {
        // error while saving job
        logger.log("Error while saving job: " + err)
      });
      this.completeCallback();
    }
  }
});

module.exports = Job;
